### PR TITLE
fix(header/p2p): add peerScore decreasing in case of failed request

### PR DESCRIPTION
**Pull request by vgonkivs**
## Overview
Resolves https://github.com/celestiaorg/celestia-node/issues/1607
## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: header/p2p/exchange_test.go [hunk 1]
```diff
@@ -322,3 +349,56 @@ func createP2PExAndServer(t *testing.T, host, tpeer libhost.Host) (*Exchange, *h
...
+	// TODO implement me
+	panic("implement me")
+}
+
+func (t *timeoutedStore) Has(context.Context, tmbytes.HexBytes) (bool, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (t *timeoutedStore) Append(context.Context, ...*header.ExtendedHeader) (int, error) {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply header/p2p/exchange_test.go:401]: Not part of getter

[the remark this answers is not in the document]
Comment (vgonkivs) [reply header/p2p/exchange_test.go]: already removed. https://github.com/celestiaorg/celestia-node/pull/1603/commits/f3167ed9dd202aef724369359d06f9588e4987d0

File: header/p2p/options.go [hunk 2]
```diff
@@ -82,6 +90,19 @@ func WithMaxRequestSize[T parameters](size uint64) Option[T] {
...
 
+// WithRequestDuration is a functional option that configures the
+// `RequestDuration` parameter.
+func WithRequestDuration[T parameters](duration time.Duration) Option[T] {
+	return func(p *T) {
+		switch t := any(p).(type) {
+		case *ClientParameters:
+			t.RequestDuration = duration
+		case *ServerParameters:
+			t.RequestDuration = duration
```
Comment (Wondertan) [inline header/p2p/options.go:101]: Does the server need it?

File: header/p2p/options.go [hunk 3]
```diff
@@ -108,6 +131,7 @@ func DefaultClientParameters() ClientParameters {
 		MaxHeadersPerRequest: 64,
 		MaxAwaitingTime:      time.Hour,
 		DefaultScore:         1,
+		RequestDuration:      time.Second * 3,
```
Comment (Wondertan) [inline header/p2p/options.go:134]: Should we make it slightly more relaxed, like 5 secs?

File: header/p2p/peer_stats.go [hunk 4]
```diff
@@ -40,6 +40,16 @@ func (p *peerStat) updateStats(amount uint64, time uint64) {
 	p.peerScore = (p.peerScore + averageSpeed) / 2
 }
 
+// decreaseScore decreases peerScore by 20% of the peer that failed the request by any reason.
+// NOTE: decreasing peerScore in one session will not affect its position in queue in another
+// session(as we can have multiple sessions running concurrently).
```
Comment (Wondertan) [inline header/p2p/peer_stats.go:45]: What is your thinking about 20%?

**Review by whatsa4 — APPROVED** [review]

File: header/p2p/options.go (hunk 2 above)
Comment (vgonkivs) [reply header/p2p/options.go:101]: to limit requests in store

File: header/p2p/peer_stats.go (hunk 4 above)
Comment (vgonkivs) [reply header/p2p/peer_stats.go:45]: We need some random value. Decided to make it 20%.

File: header/p2p/options.go (hunk 3 above)
Comment (vgonkivs) [reply header/p2p/options.go:134]: let's make 5 sec.

File: header/p2p/exchange_test.go [hunk 5]
```diff
@@ -293,6 +294,32 @@ func TestExchange_RequestByHashFails(t *testing.T) {
...
+		serverSideEx.Stop(context.Background()) //nolint:errcheck
+	})
+	exchg.peerTracker.peerLk.Lock()
+	exchg.peerTracker.trackedPeers[hosts[2].ID()] = &peerStat{peerID: hosts[2].ID(), peerScore: 200}
+	exchg.peerTracker.peerLk.Unlock()
+	_, err = exchg.GetRangeByHeight(context.Background(), 1, 3)
+	require.NoError(t, err)
+	// ensure that peerScore for the first peer was decrease by 20%
+	newPeerScore := exchg.peerTracker.trackedPeers[hosts[2].ID()].score()
+	require.NotEqual(t, 160, newPeerScore)
```
Comment (renaynay) [inline header/p2p/exchange_test.go:320]: you can do "assert.Lessthan` I think

File: header/p2p/peer_stats.go (hunk 4 above)
Comment (Wondertan) [reply header/p2p/peer_stats.go:45]: Issue

File: header/p2p/options.go (hunk 2 above)
Comment (Wondertan) [reply header/p2p/options.go:101]: https://github.com/celestiaorg/celestia-node/pull/1588

Comment (Wondertan) [reply header/p2p/options.go:101]: Isn't this PR solves it?

File: header/p2p/session.go [hunk 6]
```diff
@@ -130,17 +134,17 @@ func (s *session) doRequest(
...
 	r, size, duration, err := sendMessage(ctx, s.host, stat.peerID, s.protocolID, req)
 	if err != nil {
-		if err == context.Canceled || err == context.DeadlineExceeded {
-			return
-		}
 		log.Errorw("requesting headers from peer failed.", "failed peer", stat.peerID, "err", err)
 		select {
-		case <-ctx.Done():
 		case <-s.ctx.Done():
 			// retry request
```
Comment (renaynay) [inline header/p2p/session.go:145]: I don't understand how this leads to a retry of the request?

File: header/p2p/options.go [hunk 7]
```diff
@@ -23,14 +23,18 @@ type ServerParameters struct {
 	ReadDeadline time.Duration
 	// MaxRequestSize defines the max amount of headers that can be handled at once.
 	MaxRequestSize uint64
+	// RequestDuration defines a timeout after which session will try to re-request headers
+	// from the another peer.
```
Comment (distractedm1nd) [inline header/p2p/options.go:27]:
```suggestion
	// RequestDuration defines a timeout after which the session will try to re-request headers
	// from another peer.
```

File: header/p2p/options.go [hunk 8]
```diff
@@ -44,6 +48,10 @@ func (p *ServerParameters) Validate() error {
 	if p.MaxRequestSize == 0 {
 		return fmt.Errorf("invalid max request size: %d", p.MaxRequestSize)
 	}
+	if p.RequestDuration == 0 {
+		return fmt.Errorf("invalid RequestDuration for session: "+
```
Comment (distractedm1nd) [inline header/p2p/options.go:52]:
```suggestion
		return fmt.Errorf("invalid request duration for session: "+
```

just more consistent with the other err messages

**Review by distractedm1nd — CHANGES_REQUESTED** [review]
have the same question as rene, otherwise just comment nits

File: header/p2p/session.go (hunk 6 above)
Comment (vgonkivs) [reply header/p2p/session.go:145]: as ctx.Done is removed, we will fall in `case s.reqCh <- req:` (s.reqCh is a buffered channel, so we will not block there)

File: header/p2p/exchange_test.go [hunk 9]
```diff
@@ -322,3 +349,36 @@ func createP2PExAndServer(t *testing.T, host, tpeer libhost.Host) (*Exchange, *h
...
+	panic("implement me")
+}
+
+func (t *timeoutedStore) GetByHeight(context.Context, uint64) (*header.ExtendedHeader, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (t *timeoutedStore) GetRangeByHeight(_ context.Context, _, _ uint64) ([]*header.ExtendedHeader, error) {
+	time.Sleep(t.timeout)
```
[the remark this answers is not in the document]
Comment (vgonkivs) [reply header/p2p/exchange_test.go]: yes

File: header/p2p/session.go (hunk 6 above)
Comment (renaynay) [reply header/p2p/session.go:145]: the comment is confusing then, let's move it

**Review by renaynay — DISMISSED** [review]

**Review by renaynay — DISMISSED** [review]

File: header/p2p/options.go [hunk 10]
```diff
@@ -112,7 +116,9 @@ type ClientParameters struct {
 	MaxAwaitingTime time.Duration
 	// DefaultScore specifies the score for newly connected peers.
 	DefaultScore float32
-
+	// RequestDuration defines a timeout after which the session will try to re-request headers
+	// from another peer.
+	RequestDuration time.Duration
```
Comment (Wondertan) [inline header/p2p/options.go:121]: Let's call it timeou, duration is too general. RequestTimeout if we want to unify it for server and client.

File: header/p2p/peer_stats.go (hunk 4 above)
Comment (Wondertan) [reply header/p2p/peer_stats.go:45]: Pls

File: header/p2p/options.go (hunk 10 above)
Comment (Wondertan) [reply header/p2p/options.go:121]: Cc @renaynay, where we called it ServeTimeout

Comment (Wondertan) [reply header/p2p/options.go:121]: But RequestTimeout has more sense if we unify it

**Review by renaynay — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

**Review by Wondertan — APPROVED** [review]

**Pull request merged**